walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -117,4 +95,66 @@
     return timeRanges;
   }
 
+  /**
+   * Returns a list containing all the TimeRanges that the given collection of attendees are available for given
+   * a collection of their events for that day.
+   */
+  private List<TimeRange> determineAvailableTime(Collection<Event> events, Collection<String> attendees, long meetingDuration) {
+    List<TimeRange> availableTime = new ArrayList<>();
+    Iterator<TimeRange> unavailableTimeIterator = determineUnavailableTime(events, attendees).iterator();
+    TimeRange currentEvent = (unavailableTimeIterator.hasNext()) ?  unavailableTimeIterator.next() : null;
+    if (currentEvent != null) {
+      if (TimeRange.START_OF_DAY < currentEvent.start()
+          && currentEvent.start() - TimeRange.START_OF_DAY >= meetingDuration) {
+        availableTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.start(), false));
+      }
+
+      while (unavailableTimeIterator.hasNext()) {
+        TimeRange nextEvent = unavailableTimeIterator.next();
+
+        if (nextEvent.start() - currentEvent.end() >= meetingDuration) {
+          availableTime.add(TimeRange.fromStartEnd(currentEvent.end(), nextEvent.start(), false));
+        }
+        currentEvent = nextEvent;
+      }
+
+      if (currentEvent.end() < TimeRange.END_OF_DAY
+          && TimeRange.END_OF_DAY - currentEvent.end() + 1 >= meetingDuration) {
+        availableTime.add(TimeRange.fromStartEnd(currentEvent.end(), TimeRange.END_OF_DAY, true));
+      }
+    } else {  // There are no times taken in the day.
+      availableTime.add(TimeRange.WHOLE_DAY);
+    }
+    
+    return availableTime;
+  }
+
+  /**
+   * Given the set of meeting times, this method refits the potential meeting times to fit
+   * optional attendees with the schedule, if applicable.
+   */
+  private Collection<TimeRange> considerOptionals(
+      List<TimeRange> meetingTimes, Collection<Event> events, MeetingRequest request) {
+    if (request.getOptionalAttendees().isEmpty()) {
+      return meetingTimes;
+    }
+
+    Collection<String> allAttendees = new HashSet<>();;
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    Collection<TimeRange> totalAvailableTime = 
+        determineAvailableTime(events, allAttendees, request.getDuration());
+    Iterator<TimeRange> totalAvailableTimeIterator = totalAvailableTime.iterator();
+    
+    if (!totalAvailableTimeIterator.hasNext()) {
+      if (request.getAttendees().isEmpty()) {  // Consider when there are no mandatory attendees.
+        return totalAvailableTime;
```

Comment:
Is this the desired behavior? The way this is currently coded, if there are no mandatory attendees and no way to fit in all the optional attendees you will return the empty list.  My understanding, though, is that you should return all time slots because any time range is valid if there are no mandatory attendees.  I think removing this if case entirely would be the correct behavior.

Reply:
Yes, that's the desired behavior. One of the test cases provided states that if there are no mandatory attendees, and there are two optional attendees who have no gaps in their schedules, then the result should be that there are no available times at all.

Reply:
Yes, I see that now.  It makes no sense to me but  I guess that's how this project wants you to handle it.